### DISCUSSION TITLE
Identification of WS28xx-like protocol

### DISCUSSION THREAD

**Discussion by scy** [Makuna/NeoPixelBus#819] [Q&A] [answered]
Hi there! I have bought a RGB LED curtain ([this one](https://www.amazon.de/dp/B0D2CVN69C), "Traumheim Smart Color Changing Lights") and noticed that it doesn't quite work in WLED, no matter which protocol I choose.

So I pulled out the scope and attached it to the original controller, and found out that the timings are out of spec for WS2812. Instead, they are:

* **T0H:** 330 ns
* **T0L:** 1080 ns
* **T1H:** 1170 ns
* **T1L:** 660 ns

These values appear to be pretty constant, and yes, the symbol lengths significantly differ between a 0 (1410 ns) and a 1 (1830 ns).

I'm not a C++ programmer, but I can talk to that strip without any issues using MicroPython with a custom NeoPixel timing:

```python
leds = NeoPixel(pin, 400, timing=(330, 1080, 1170, 660))
```

The curtain uses 20 strands of 20 LEDs, with each LED enclosed in a white ball around 15 mm in diameter, rather hard (but still somewhat squishy) silicone or plastic. Color order on the wire appears to be RGB, not GRB. I don't know whether the LEDs use fixed addresses or whether there's some kind of chip in the small white plastic housing on the top of each strand.

I don't know whether NeoPixelBus supports this protocol already because I'm quite confused by the plethora of WS28xx variants. Can anyone of you please tell me which protocol this is, so I can possibly talk to the WLED devs to add support for it?

**Comment by Makuna** [comment]
You didn't mention which platforms you use.  Esp32?  Esp8266?

Some "methods" (how we use hardware to generate the signals) can't support asymmetrical timing like that.  But often specs are not very accurate.

With your working micro python, try these and see if they work
(330, 1080, 1080, 330)
(300, 950, 900, 350) // ws2811
(300, 900, 900, 300) // GS1903

**Reply by scy** [reply]
Thanks for getting back to me that fast! I'm currently using an ESP32.

All three timings you provided seem to work, but my test isn't very sophisticated. (It's basically just "fill the strip with one color").

I was thinking about hacking together a simple [DDP](http://www.3waylabs.com/ddp/) implementation for MicroPython so that I can have WLED control the curtain via the network and display its builtin effects on it. I'll report back.

**Reply by Makuna** [reply]
If all those tests work, then the last two should be selectable in WLED.  Look for WS2811 and GS1903 and try them with WLED.

**Reply by scy** [reply]
My DDP server is working and I was able to have WLED stream pixel data via UDP to my MicroPython code and try out the three timings you've suggested.

All of them work without any issues. :star_struck: (When using my MicroPython solution and its timings.)

However, WLED neither has WS2811 nor GS1903 available in the protocol selection dropdown. There's WS281x, SK6812/WS2814 RGBW, TM1814, 400kHz, TM1829, UCS8903, APA106/PL9823, TM1914, and a bunch of RGB(C)W and PWM ones as well as some that require a clock line.

From the available options, "400kHz" (which is apparently WS2811 at 400 kHz, not 800) _almost_ works, but there's random flickering of single LEDs.

But there is another possibility: That curtain has originally had almost 5 m of (my estimate) 0.3 mm² (22 AWG) wire between the controller and the first LED. This is how the signal looked when I attached my scope close to the controller, with the whole curtain connected:

![image](https://github.com/user-attachments/assets/ed6e52c4-64ea-47c0-b19c-cd33beba82ce)

Without the curtain connected, it looked like this:

![image](https://github.com/user-attachments/assets/5ccef461-65a5-4565-bbf8-6ec3c6a517bf)

Could it be that the original controller is using fine-tuned out-of-spec timings to make up for the effects of the very long wire?

I could try cutting the strip significantly closer to the first LED and check whether it runs with standard timings then, but I'd have to spend some extra effort in making it waterproof again.

Any thoughts? And thanks again for the support so far!

**Reply by scy** [reply, marked as answer]
I've cut the wire down from almost 5 m to just 3 m. I can now successfully talk to the LEDs using WLED's "400 kHz" protocol.

Thanks again for your efforts @Makuna, I highly appreciate it!

**Reply by Makuna** [reply]
You are seeing the results of resistance, impendence, and capacitance of the circuit (uC, wires, and LEDs).  
If you search around, you will see where people talk about needing a resistor and or capacitor.  These are added to tune the circuit to provide a cleaning signal.  Its magic, black magic 8-)
With the scope, you can play with the values of the resistor and or capacitor.
Also note, the Esp32 is 3.3v, while the original controller and LEDs look like 5v.  Use a good level converter can help also here if you need the extra meters of distance.
